ably/realtime_conn.go

```diff
@@ -480,44 +531,50 @@ func (c *Connection) eventloop() {
 			c.auth.updateClientID(msg.ConnectionDetails.ClientID)
 			if msg.ConnectionDetails != nil {
 				c.state.Lock()
-				c.details = *msg.ConnectionDetails
+				c.key = msg.ConnectionDetails.ConnectionKey //(RTN15e) (RTN16d)
 				c.state.Unlock()
 
 				// Spec RSA7b3, RSA7b4, RSA12a
-				c.auth.updateClientID(c.details.ClientID)
+				c.auth.updateClientID(msg.ConnectionDetails.ClientID)
 
 				maxIdleInterval := time.Duration(msg.ConnectionDetails.MaxIdleInterval) * time.Millisecond
 				receiveTimeout = c.opts.realtimeRequestTimeout() + maxIdleInterval // RTN23a
 			}
 			c.state.Lock()
-			reconnecting := c.reconnecting
-			if reconnecting {
-				c.reconnecting = false
+			mode := c.mode
+			if mode != NormalMode {
+				// reset the mode
+				c.mode = NormalMode
 			}
+			id := c.id
 			c.state.Unlock()
-			if reconnecting {
-				// (RTN15c1) (RTN15c2)
+			switch mode {
+			case ResumeMode, RecoveryMode:
+				// (RTN15c1) (RTN15c2) (RTN16e)
 				c.state.Lock()
 				c.setState(StateConnConnected, newErrorProto(msg.Error))
-				id := c.id
 				c.state.Unlock()
 				if id != msg.ConnectionID {
-					// (RTN15c3)
-					// we are calling this outside of locks to avoid deadlock because in the
-					// RealtimeClient client where this callback is implemented we do some ops
-					// with this Conn where we re acquire Conn.state.Lock again.
-					c.callbacks.onReconnectMsg(msg)
+					// (RTN16a) its up to client to explicitly attach channels when in
+					// RecoveryMode
+					if mode == ResumeMode {
```

Comment:
Note that if we're recovering instead of resuming, we can also call `onReconnectMsg` anyway since there won't be any channels to reattach, so the behavior is the same. Given this, the only bit of logic that diverges here between `ResumeMode` and `RecoveryMode` disappears, so we can just go back to the `reconnecting` boolean flag.

Reply:
addressed in 4449023b755bb04b26b63f4dab7c83ccd5dba667